### DISCUSSION TITLE
Modern library authoring

### DISCUSSION THREAD

**Discussion by vanillajonathan** [dotnet/runtime#84928] [Ideas]
If you're writing a library you're probably targeting `netstandard2.0` which can be quite frustrating since you cannot use modern features such as implicit usings, file-scoped namespaces or nullable references types.

Please improve the library authoring developer experience by supporting modern features and development practices such as implicit usings, file-scoped namespaces and nullable references types.

**Comment by teo-tsirpanis** [comment]
Hello! C# versions larger than 7.3 are not officially supported in .NET Standard 2.0. However you can opt in to the latest language version by adding the `<LangVersion>latest</LangVersion>` property in your project file.

**Reply by vanillajonathan** [reply]
Yes, I am aware of that. I would like to see official support for newer versions of C# in .NET Standard 2.0 and a better out-of-the-box experience when scaffolding a new library project.

**Reply by teo-tsirpanis** [reply]
I don't think this is going to happen. New language versions depend on attributes that don't get backported to earlier frameworks.

**Reply by vanillajonathan** [reply]
Yeah, I can understand not everything can be backported (such as records, init-setters, etc), but file-scoped namespaces and implicit using sounds like something that maybe could be backported. Either way, I think Microsoft someway, somehow should improve the library authoring process because it feels outdated.

**Comment by huoyaoyuan** [comment]
You can use libraries like [PolySharp](https://www.nuget.org/packages/PolySharp) to fulfill some attributes for down level frameworks.
For me, I'm dropping .NET Framework and .NET Standard in my libraries when more feature and scenarios are supported by .NET Core.

**Reply by vanillajonathan** [reply]
Then what do you target and how do you distribute such a library? You going to target `net6.0` or multi-target `net5.0;net6.0;net7.0;net8.0`?

**Reply by huoyaoyuan** [reply]
A reasonable lowest like `net5.0` or `net6.0`, plus the current version.

**Reply by vanillajonathan** [reply]
But does that work across all the .NET versions between the lowest and the current? Or do I have to add a target moniker for every version between the lowest and the current version?

**Reply by huoyaoyuan** [reply]
> Or do I have to add a target moniker for every version between the lowest and the current version?

Every version between them will resolve to the lower version as a fallback.
